File: C#/EventsTest/Form1.cs
using Cybele.Thinfinity;
using System;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Windows.Forms;

namespace VirtualUIEvents
{
    public partial class Form1 : Form {
        private VirtualUI Vui;

        private string AppPath;
        private string m_TestDir;
        private string m_LogfileName;

        private void PrepareTestDir() {
            m_TestDir = Path.Combine(AppPath, "test_" + Process.GetCurrentProcess().Id.ToString());
            LblTestDir.Text = "Test directory: " + m_TestDir;

            // Create test directory:
            Directory.CreateDirectory(m_TestDir);

            // Create subdir for uploads:
            Directory.CreateDirectory(Path.Combine(m_TestDir, "Uploads"));

            // Create text file to be downloaded:
            StreamWriter file = new StreamWriter(Path.Combine(m_TestDir, "test.txt"));
            file.WriteLine("File to download by PID " + Process.GetCurrentProcess().Id.ToString());
            file.Close();
        }

        private void RemoveTestDir() {
            if (!String.IsNullOrEmpty(m_TestDir)) {
                Directory.Delete(m_TestDir, true);
            }
        }

        private void Log(string data) {
            TxtLog.AppendText(data + "\r\n");
            if (CheckSaveLog.Checked) {
                StreamWriter file = new StreamWriter(m_LogfileName, true);
                file.WriteLine(data);
                file.Close();
            }
        }

        public Form1() {
            InitializeComponent();
            AppPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().CodeBase);
            if (AppPath.StartsWith("file:")) {
                AppPath = AppPath.Substring(6);
            }
            
            Vui = new VirtualUI();
            Vui.Start();
            Vui.AllowExecute(".+");

            //Deprecated
            Vui.OnDragFile += Vui_OnDragFile;
            Vui.OnDragFile2 += Vui_OnDragFile2;
            Vui.OnDropFile += Vui_OnDropFile;
            Vui.OnDownloadEnd += Vui_OnDownloadEnd;
            Vui.OnUploadEnd += Vui_OnUploadEnd;
            Vui.OnGetUploadDir += Vui_OnGetUploadDir;
            Vui.OnClose += Vui_OnClose;
            Vui.OnBrowserResize += Vui_OnBrowserResize;
            Vui.OnSaveDialog += Vui_OnSaveDialog;

            // Currently not tested on this app:
            Vui.OnReceiveMessage += Vui_OnReceiveMessage;
            Vui.OnRecorderChanged += Vui_OnRecorderChanged;




            Vui.BrowserInfo.SetCookie("test2", "Hello2", "");

            GetCookieLabel.Text = Vui.BrowserInfo.GetCookie("test1");
            GetCookieLabel2.Text = Vui.BrowserInfo.GetCookie("test2");


            PrepareTestDir();
            m_LogfileName = Path.Combine(AppPath, String.Format("VirtualUI_Events_{0}.txt", Process.GetCurrentProcess().Id));
            CheckSaveLog.Text = "Save log to " + Path.GetFileName(m_LogfileName);
        }

        private void Form1_FormClosed(object sender, FormClosedEventArgs e) {
            if (CheckRemoveTestDir.Checked) {
                RemoveTestDir();
            }
        }

        private void BtnDownload_Click(object sender, EventArgs e) {
            Vui.DownloadFile(Path.Combine(m_TestDir, "test.txt"));
        }

        private void button1_Click(object sender, EventArgs e)
        {
            
            SaveFileDialog saveFileDialog1 = new SaveFileDialog();


            saveFileDialog1.Title = ".NET Open Dialog";
            saveFileDialog1.Filter = "Image files (*.png;*.jpeg;*.bmp)|*.png;*.jpg;*.bmp|Text files (*.txt;*.log;*.ini)|*.txt;*.log;*.ini|PDF Files (*.pdf)|*.pdf ";
            saveFileDialog1.FilterIndex = 3;
            saveFileDialog1.FileName = "sample.txt";
            saveFileDialog1.DefaultExt = "txt";
            saveFileDialog1.ShowHelp = true;
            //saveFileDialog1.ShowReadOnly = true;
            saveFileDialog1.InitialDirectory = ".\\";
            saveFileDialog1.SupportMultiDottedExtensions = true;
            
           

            saveFileDialog1.ShowDialog();

            
            //StreamWriter PDFfile = new StreamWriter(saveFileDialog1.FileName);

            StreamWriter Test = new StreamWriter(Path.Combine(m_TestDir, saveFileDialog1.FileName));
            Test.WriteLine("File to download by PID " + Process.GetCurrentProcess().Id.ToString());
            Test.Close();


            Vui.DownloadFile(saveFileDialog1.FileName);
            

            

        }

        private void BtnUpload_Click(object sender, EventArgs e) {

            OpenFileDialog openFileDialog1 = new OpenFileDialog();

            openFileDialog1.Title = ".NET Open Dialog";
            openFileDialog1.Filter = "Image files (*.png;*.jpeg;*.bmp)|*.png;*.jpg;*.bmp|Text files (*.txt;*.log;*.ini)|*.txt;*.log;*.ini|PDF Files (*.pdf)|*.pdf ";
            openFileDialog1.FilterIndex = 3;
            openFileDialog1.FileName = "test.pdf";
            openFileDialog1.DefaultExt = "txt";

            openFileDialog1.InitialDirectory = ".\\";

            openFileDialog1.Multiselect = true;


            //openFileDialog1.ShowHelp = true;
            //openFileDialog1.SupportMultiDottedExtensions = true;


            openFileDialog1.ShowDialog();

            //Vui.UploadFile();
        }

        private void Vui_OnSaveDialog(object sender, SaveDialogArgs e) {
            Log("OnSaveDialog: " + e.Filename);
        }

        private void Vui_OnRecorderChanged(object sender, RecorderChangedArgs e) {
            Log("OnRecorderChanged");
        }

        private void Vui_OnReceiveMessage(object sender, ReceiveMessageArgs e) {
            Log("OnReceiveMessage: " + e.Data);
        }

        private void Vui_OnBrowserResize(object sender, BrowserResizeEventArgs e) {
            Log(String.Format("OnBrowserResize: {0}x{1}", e.Width, e.Height));
        }

        private void Vui_OnClose(object sender, CloseArgs e) {
            Log("OnClose");
            Close();
        }

        private void Vui_OnGetUploadDir(object sender, GetUploadDirEventArgs e) {
            e.Directory = Path.Combine(m_TestDir, "Uploads");
            e.Handled = true;
            Log("OnGetUploadDir: Set " + e.Directory);
        }

        private void Vui_OnUploadEnd(object sender, UploadEndArgs e) {
            Log("OnUploadEnd: " + e.Filename);
        }

        private void Vui_OnDownloadEnd(object sender, DownloadEndArgs e) {
            Log("OnDownloadEnd: " + e.Filename);
        }

        private void Vui_OnDragFile(object sender, DragFileArgs e) {
            // Using OnDragfile2
        }

        private void Vui_OnDragFile2(object sender, DragFile2Args e) {
            string actionStr = "Unknown";
            switch (e.Action) {
                case Cybele.Thinfinity.DragAction.Start:
                    actionStr = "Start";
                    break;
                case Cybele.Thinfinity.DragAction.Over:
                    actionStr = "Over";
                    break;
                case Cybele.Thinfinity.DragAction.Drop:
                    actionStr = "Drop";
                    break;
                case Cybele.Thinfinity.DragAction.Error:
                    actionStr = "Error";
                    break;
                case Cybele.Thinfinity.DragAction.Cancel:
                    actionStr = "Cancel";
                    break;
            }

            string eventData = String.Format("{0} on {1}.{2}", actionStr, e.ScreenX, e.ScreenY);
            if (!String.IsNullOrEmpty(e.Filenames)) {
                eventData += " (Files: " + e.Filenames + ")";
            }
            Log("OnDragFile2: " + eventData);

            e.Accept = true;
        }

        private void Vui_OnDropFile(object sender, DropFileArgs e) {
            string eventData = String.Format("{0}.{1}", e.ScreenX, e.ScreenY);
            if (!String.IsNullOrEmpty(e.Filenames)) {
                eventData += " (Files: " + e.Filenames + ")";
            }
            if (!String.IsNullOrEmpty(e.FileSizes)) {
                eventData += " (Sizes: " + e.FileSizes + ")";
            }
            Log("OnDropFile: " + eventData);

            //TODO: Test IgnoreFiles (Set in the form: "ignore1.txt|ignore2.log|ignoreN.exe")
            Vui.UploadFile();
        }


    }
}

File: VB.NET/EventsTest/Form1.vb
Imports System.IO
Imports System.Reflection

Public Class Form1
    Private Vui As Cybele.Thinfinity.VirtualUI

    Private AppPath As String
    Private m_TestDir As String
    Private m_LogfileName As String

    Private Sub PrepareTestDir()
        m_TestDir = Path.Combine(AppPath, "test_" + Process.GetCurrentProcess.Id.ToString())
        LblTestDir.Text = "Test directory: " + m_TestDir

        ' Create test directory:
        Directory.CreateDirectory(m_TestDir)

        ' Create subdir for uploads:
        Directory.CreateDirectory(Path.Combine(m_TestDir, "Uploads"))

        ' Create text file to be downloaded:
        Dim file As StreamWriter = New StreamWriter(Path.Combine(m_TestDir, "test.txt"))
        file.WriteLine("File to download by PID " + Process.GetCurrentProcess.Id.ToString())
        file.Close()
    End Sub

    Private Sub RemoveTestDir()
        If (Not String.IsNullOrEmpty(m_TestDir)) Then
            Directory.Delete(m_TestDir, True)
        End If
    End Sub

    Private Sub Log(data As String)
        TxtLog.AppendText(data + vbCrLf)
        If (CheckSaveLog.Checked) Then
            Dim file As StreamWriter = New StreamWriter(m_LogfileName, True)
            file.WriteLine(data)
            file.Close()
        End If
    End Sub


    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        AppPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().CodeBase)
        If (AppPath.StartsWith("file:")) Then
            AppPath = AppPath.Substring(6)
        End If

        Vui = New Cybele.Thinfinity.VirtualUI()
        Vui.Start(60)
        AddHandler Vui.OnDownloadEnd, AddressOf Vui_OnDownloadEnd
        AddHandler Vui.OnUploadEnd, AddressOf Vui_OnUploadEnd
        AddHandler Vui.OnGetUploadDir, AddressOf Vui_OnGetUploadDir
        AddHandler Vui.OnClose, AddressOf Vui_OnClose
        AddHandler Vui.OnBrowserResize, AddressOf Vui_OnBrowserResize
        AddHandler Vui.OnReceiveMessage, AddressOf Vui_OnReceiveMessage
        AddHandler Vui.OnDragFile, AddressOf Vui_OnDragFile
        AddHandler Vui.OnDragFile2, AddressOf Vui_OnDragFile2
        AddHandler Vui.OnDropFile, AddressOf Vui_OnDropFile

        ' Currently not tested on this app:
        AddHandler Vui.OnRecorderChanged, AddressOf Vui_OnRecorderChanged
        AddHandler Vui.OnSaveDialog, AddressOf Vui_OnSaveDialog

        PrepareTestDir()
        m_LogfileName = Path.Combine(AppPath, String.Format("VirtualUI_Events_{0}.txt", Process.GetCurrentProcess.Id))
        CheckSaveLog.Text = "Save log to " + Path.GetFileName(m_LogfileName)
    End Sub

    Private Sub Form1_FormClosed(sender As Object, e As FormClosedEventArgs) Handles MyBase.FormClosed
        If (CheckRemoveTestDir.Checked) Then
            RemoveTestDir()
        End If
    End Sub

    Private Sub ButtonDownload_Click(sender As Object, e As EventArgs) Handles ButtonDownload.Click
        Vui.DownloadFile(Path.Combine(m_TestDir, "test.txt"))
    End Sub

    Private Sub ButtonUpload_Click(sender As Object, e As EventArgs) Handles ButtonUpload.Click
        Vui.UploadFile()
    End Sub


    Private Sub Vui_OnBrowserResize(sender As Object, e As Cybele.Thinfinity.BrowserResizeEventArgs)
        Log(String.Format("OnBrowserResize: {0}x{1}", e.Width, e.Height))
    End Sub

    Private Sub Vui_OnClose(Sender As Object, e As Cybele.Thinfinity.CloseArgs)
        Log("OnClose")
        Close()
    End Sub

    Private Sub Vui_OnDownloadEnd(Sender As Object, e As Cybele.Thinfinity.DownloadEndArgs)
        Log("OnDownloadEnd: " + e.Filename)
    End Sub

    Private Sub Vui_OnGetUploadDir(Sender As Object, e As Cybele.Thinfinity.GetUploadDirEventArgs)
        e.Directory = Path.Combine(m_TestDir, "Uploads")
        e.Handled = True
        Log("OnGetUploadDir: Set " + e.Directory)
    End Sub

    Private Sub Vui_OnUploadEnd(Sender As Object, e As Cybele.Thinfinity.UploadEndArgs)
        Log("OnUploadEnd: " + e.Filename)
    End Sub

    Private Sub Vui_OnSaveDialog(Sender As Object, e As Cybele.Thinfinity.OnSaveDialogArgs)
        Log("OnSaveDialog: " + e.Filename)
    End Sub

    Private Sub Vui_OnDragFile(Sender As Object, e As Cybele.Thinfinity.DragFileArgs)
        ' Using OnDragFile2
    End Sub

    Private Sub Vui_OnDragFile2(Sender As Object, e As Cybele.Thinfinity.DragFile2Args)
        Dim actionStr As String = "Unknown"
        Select Case e.Action
            Case Cybele.Thinfinity.DragAction.DRAG_Start
                actionStr = "Start"
            Case Cybele.Thinfinity.DragAction.DRAG_Over
                actionStr = "Over"
            Case Cybele.Thinfinity.DragAction.DRAG_Drop
                actionStr = "Drop"
            Case Cybele.Thinfinity.DragAction.DRAG_Error
                actionStr = "Error"
            Case Cybele.Thinfinity.DragAction.DRAG_Cancel
                actionStr = "Cancel"
        End Select

        Dim eventData As String = String.Format("{0} on {1}.{2}", actionStr, e.ScreenX, e.ScreenY)
        If (Not String.IsNullOrEmpty(e.Filenames)) Then
            eventData = eventData + " (Files: " + e.Filenames + ")"
        End If
        Log("OnDragFile2:" + eventData)

        e.Accept = True
    End Sub

    Private Sub Vui_OnDropFile(Sender As Object, e As Cybele.Thinfinity.DropFileArgs)
        Dim eventData As String = String.Format("{0}.{1}", e.ScreenX, e.ScreenY)
        If (Not String.IsNullOrEmpty(e.Filenames)) Then
            eventData = eventData + " (Files: " + e.Filenames + ")"
        End If
        If (Not String.IsNullOrEmpty(e.FileSizes)) Then
            eventData = eventData + " (Sizes: " + e.FileSizes + ")"
        End If
        Log("OnDropFile: " + eventData)

        'TODO: Test IgnoreFiles (Set in the form: "ignore1.txt|ignore2.log|ignoreN.exe")
        Vui.UploadFile()
    End Sub

    Private Sub Vui_OnReceiveMessage(Sender As Object, e As Cybele.Thinfinity.ReceiveMessageArgs)
        Log("OnReceiveMessage: " + e.Data)
    End Sub

    Private Sub Vui_OnRecorderChanged(Sender As Object, e As Cybele.Thinfinity.RecorderChangedArgs)
        Log("OnRecorderChanged")
    End Sub
End Class
